plugins/image-prioritizer/class-ip-img-tag-visitor.php

```diff
@@ -0,0 +1,99 @@
+<?php
+/**
+ * Image Prioritizer: IP_Img_Tag_Visitor class
+ *
+ * @package image-prioritizer
+ * @since n.e.x.t
+ */
+
+// Exit if accessed directly.
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Visitor for the tag walker that optimizes IMG tags.
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+final class IP_Img_Tag_Visitor extends IP_Tag_Visitor {
+
+	/**
+	 * Visits a tag.
+	 *
+	 * @param OD_HTML_Tag_Walker $walker Walker.
+	 * @return bool Whether the visitor visited the tag.
+	 */
+	public function __invoke( OD_HTML_Tag_Walker $walker ): bool {
+		if ( 'IMG' !== $walker->get_tag() ) {
+			return false;
+		}
+
+		// Skip empty src attributes and data: URLs.
+		$src = trim( (string) $walker->get_attribute( 'src' ) );
+		if ( '' === $src || $this->is_data_url( $src ) ) {
+			return false;
+		}
+
+		$xpath = $walker->get_xpath();
+
+		// Ensure the fetchpriority attribute is set on the element properly.
+		$common_lcp_element = $this->url_metrics_group_collection->get_common_lcp_element();
+		if ( ! is_null( $common_lcp_element ) && $xpath === $common_lcp_element['xpath'] ) {
+			if ( 'high' === $walker->get_attribute( 'fetchpriority' ) ) {
+				$walker->set_attribute( 'data-od-fetchpriority-already-added', true );
+			} else {
+				$walker->set_attribute( 'fetchpriority', 'high' );
+				$walker->set_attribute( 'data-od-added-fetchpriority', true );
+			}
+
+			// Never include loading=lazy on the LCP image common across all breakpoints.
+			if ( 'lazy' === $walker->get_attribute( 'loading' ) ) {
+				$walker->set_attribute( 'data-od-removed-loading', $walker->get_attribute( 'loading' ) );
+				$walker->remove_attribute( 'loading' );
+			}
+		} elseif ( is_string( $walker->get_attribute( 'fetchpriority' ) ) && $this->url_metrics_group_collection->is_every_group_populated() ) {
+			// Note: The $all_breakpoints_have_url_metrics condition here allows for server-side heuristics to
+			// continue to apply while waiting for all breakpoints to have metrics collected for them.
+			$walker->set_attribute( 'data-od-removed-fetchpriority', $walker->get_attribute( 'fetchpriority' ) );
+			$walker->remove_attribute( 'fetchpriority' );
```

Comment:
could we set fetchpriority to low here instead of removing?

Reply:
I see the comment here explaining the removal is not fleshed out enough. I've elaborated in 489d1f6b. It may be that the element here is actually the LCP element for _some_ of the viewport groups. In this case, `fetchpriority=high` is removed from the `img` but it is _added_ to the preload link below which also includes the media query to ensure it is only prioritized for the right viewports.

Adding `fetchpriority=low` is something I want to explore in the context of images that are above the fold but obscured, such as in a mega menu or as subsequent slides in a carousel.